商品購入機能

cony-k-m90 opened the pull request:
# What
コードレビューをお願いいたします。

# Why
商品購入機能を実装するため。

---
### 一部動作についてキャプチャーを取得しているので、ご確認ください。
- 入力に問題がある状態で購入ボタンが押されたら、購入ページに戻りエラーメッセージが表示されること
https://gyazo.com/38cf480c380c522fe10a1b4306e40f56

- 購入が完了したら、トップページまたは購入完了ページに遷移すること
https://gyazo.com/401ae86220c5aad3bd9e4a5136b038da

- ログアウト状態のユーザーは、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移すること
https://gyazo.com/b96436f7e97944d152ecf8df7e427b54
https://gyazo.com/5607c48156949d577836c50444cc6793

- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移すること
https://gyazo.com/93080202f4f0b449970777ac4c2a3dc6

- ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移すること
https://gyazo.com/d0a4c3630eccd0c5b1611b8f3b96941b

cony-k-m90 committed:
orderモデルを作成

cony-k-m90 committed:
READMEを更新

cony-k-m90 committed:
items/ordersへのルーティングを追記

「購入画面へ進む」から商品を「購入」するために必要なindexとcreateのルーティングを記述。

cony-k-m90 committed:
ordersテーブルのマイグレーション

order.rbモデルを作成していたが、マイグレーションができていなかった。

cony-k-m90 committed:
orderコントローラー作成と「購入内容の確認」ページへ遷移するルート設定

cony-k-m90 committed:
商品購入画面の表示

cony-k-m90 committed:
addressモデル作成、item_order.rbフォームオブジェクト作成

作成したモデルとオブジェクトに伴ってER図とREADME修正、モデルのカラム修正、テーブル間のリレーションを修正

cony-k-m90 committed:
orders/index.html.erbに都道府県のアクティブハッシュを反映

cony-k-m90 committed:
ordersへのルート、コントローラーに「:new」アクションを追加

cony-k-m90 committed:
order_controller.rbへストロングパラメーターの記述

cony-k-m90 committed:
「購入」ボタンを押すとテーブルにデータを保存

「orders」「addresses」各テーブルにパラメーターが保存できる状態になった。
また同じ商品を何度でも購入できてしまう。

cony-k-m90 committed:
商品購入画面のバリデーションとコントローラー修正

必須入力項目がからの場合に、同じ画面上にエラーを出力する処理ができていなかった。

cony-k-m90 committed:
エラーハンドリング、フォームオブフェクトのテストコード作成

フォームオブジェクトへ入力されたパラメーターのバリデーションの修正を行った。

cony-k-m90 committed:
"turbolinks"を使用しないように設定箇所の記述を変更

pay.jp実装前の準備

cony-k-m90 committed:
payjpから取得したトークンをサーバーサイドへ送信

JavaScript（card.js）ファイルの作成、トークンをpayjpから取得、サーバーサイドへ送信まで。

cony-k-m90 committed:
payjp決済機能を実装

テスト用だが、payjpの「公開鍵」を記述したままpushしてしまい問題があった。
今回のcommitでは、「秘密鍵」のコントローラーへの追記と、「秘密鍵」「公開鍵」を環境変数に書き換た。
また前回pushしてしまった「公開鍵」が使えない様、payjp側で「公開鍵」を更新した。

cony-k-m90 committed:
売却済みの商品に対して「sold out」の表示をする

cony-k-m90 committed:
購入済みの場合出品者が商品情報の編集を行えない様にする

cony-k-m90 committed:
商品購入ページへの遷移を制御

- ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移すること
- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移すること
- ログアウト状態のユーザーは、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移すること

cony-k-m90 committed:
商品が購入済みの場合編集できない

orderモデル(ordersテーブル)作成後、購入済みの商品が判定できる様になった。

cony-k-m90 committed:
単体テスト実施

cony-k-m90 committed:
Rubocopj実行と自動修正実行

div-reviewer reviewed (commented):
cony-k-m90様、学習お疲れ様です！
お忙しい中商品購入のレビュー依頼ありがとうございます。プルリクエスト確認させていただきました。

可読性の高い綺麗なコーディングにしようという意識が伝わるコードです。素晴らしいですね。
修正点をコメントさせて頂きましたので、ご確認をお願い致します。

```ruby
【次回プルリクエスト提出する際の注意点】
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```

ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
どうぞよろしくおねがいします。

cony-k-m90 committed:
フォームオブジェクトへ渡す必要の無いパラメータ削除(orders_controller.rb)

金額はフォームオブジェクト内に持っていく必要はないので削除しましょう。
商品出品で既にバリデーションも保存も行っている為。

cony-k-m90 committed:
不要なストロングパラメーターの削除(orders_controoller.rb)

before_action :authenticate_user!, only: [:index, :create]
で行っている作業と同じ為です。

cony-k-m90 commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,52 @@
+  def item_order_params
+    # require: バリデーションで指定した値が存在しなければ例外を出す
+    params.require(:item_order).permit(:postal_code, :shipping_area_id, :city, :house_number, :building_name, :phone_number).merge(
+      user_id: current_user.id, item_id: params[:item_id], token: params[:token], price: @item.price
```
フォームオブジェクトから、pay.jpへ渡すものと勘違いしていました。
削除完了しました。

cony-k-m90 commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,52 @@
+
+  def move_to_index
+    redirect_to root_path unless user_signed_in?
+  end
```
ありがとうございます。
authenticate_user!メソッドが使えると気付いたのですが、不要になる箇所を考えていませんでした。

cony-k-m90 commented on app/models/item_order.rb:
```diff
@@ -0,0 +1,21 @@
+  attr_accessor :postal_code, :shipping_area_id, :city, :house_number, :building_name, :phone_number, :user_id, :item_id,
+                :order_id, :token, :price
+
+  # ここに住所のバリデーションを書く
```
ありがとうございます。
orders_controller.rb 側で削除したものですね。削除しました。

cony-k-m90 commented on app/models/item_order.rb:
```diff
@@ -0,0 +1,21 @@
+    validates :shipping_area_id, numericality: { other_than: 0, message: "can't be blank" }
+    validates :city, format: { with: /\A[ぁ-んァ-ヶ一-龥々ー]+\z/, message: 'is invalid. Input full-width characters' }
+    validates :house_number, :token, :price
+    validates :phone_number, format: { with: /\A[0-9]{11}\z/ }
```
修正完了しました。

cony-k-m90 committed:
商品購入時のバリデーションを修正

- フォームオブジェクトのバリデーション
電話番号は10桁の場合もあるため、10桁、11桁登録が可能なようにしましょう！
:priceは削除しましょう。
価格は商品出品で登録する際に既にバリデーションをかけている為です。

cony-k-m90 commented on app/models/item_order.rb:
```diff
@@ -0,0 +1,21 @@
+    validates :postal_code, format: { with: /\A[0-9]{3}-[0-9]{4}\z/, message: 'is invalid. Include hyphen(-)' }
+    validates :shipping_area_id, numericality: { other_than: 0, message: "can't be blank" }
+    validates :city, format: { with: /\A[ぁ-んァ-ヶ一-龥々ー]+\z/, message: 'is invalid. Input full-width characters' }
+    validates :house_number, :token, :price
```
ありがとうございます。
確かに不要です。削除しました。

cony-k-m90 committed:
案内用文の精査

不要な案内文の削除と、残しておきたいメモの精査

cony-k-m90 commented on app/views/items/index.html.erb:
```diff
@@ -134,9 +134,11 @@
+                    </div>
+                  <% end %>
                 <%# //商品が売れていればsold outを表示しましょう %>
               </div>
```
ありがとうございます。
ほとんどの案内文は削除しましたが、説明として必要そうと感じたものは残しました。

cony-k-m90 commented on app/views/items/show.html.erb:
```diff
@@ -9,9 +9,11 @@
+        </div>
+      <% end %>
       <%# //商品が売れている場合は、sold outを表示しましょう %>
     </div>
```
ありがとうございます。
ほとんどの案内文は削除しましたが、説明として必要そうと感じたものは残しました。

cony-k-m90 commented on app/views/items/show.html.erb:
```diff
@@ -25,12 +27,16 @@
+          <%= link_to "購入画面に進む", item_orders_path(@item.id), method: :get, class:"item-red-btn"%>
+        <% end %>
       <% end %>
     <%# //商品が売れていない場合はこちらを表示しましょう %>
```
ありがとうございます。
ほとんどの案内文は削除しましたが、説明として必要そうと感じたものは残しました。

cony-k-m90 commented on app/views/items/show.html.erb:
```diff
@@ -25,12 +27,16 @@
+        <% unless @item.order.present? %>
+          <%= link_to "購入画面に進む", item_orders_path(@item.id), method: :get, class:"item-red-btn"%>
+        <% end %>
       <% end %>
```
ありがとうございます。
blank? メソッドを使うことで解決できました。

cony-k-m90 committed:
可読性が悪いためリファクタリング

共通する条件を上位にして、分岐させる条件から削除した。

cony-k-m90 commented on app/controllers/items_controller.rb:
```diff
@@ -23,7 +23,11 @@ def show
 
   def edit
-    redirect_to action: :index unless @item.user == current_user
+    if @item.order.present?
+      redirect_to action: :index
+    else
+      redirect_to action: :index unless @item.user == current_user
+    end
```
ありがとうございます。
無理やり後から条件を追加したので、この様な形になっていました。
今回の場合ですと、商品詳細画面の編集ができる条件を言葉で整理してから、コードに置き換えるとどうなるかを考えるとわかりやすかったです。
シンプルにできました。

cony-k-m90 commented on spec/models/item_order_spec.rb:
```diff
@@ -0,0 +1,103 @@
+    end
+
+    context '「クレジットカード情報入力」の項目に不備があると決済（商品購入）ができないこと' do
+      it 'priceが空では保存ができないこと' do
+        @item_order.price = nil
+        @item_order.valid?
+        expect(@item_order.errors.full_messages).to include("Price can't be blank")
+      end
```
ありがとうございます。
削除しました。

cony-k-m90 commented on spec/models/item_order_spec.rb:
```diff
@@ -0,0 +1,103 @@
+      it 'postal_code が空だと保存できないこと' do
+        @item_order.postal_code = ''
+        @item_order.valid?
+        expect(@item_order.errors.full_messages).to include("Postal code can't be blank",
```
ありがとうございます。
不要なメッセージを削除しました。

cony-k-m90 commented on spec/models/item_order_spec.rb:
```diff
@@ -0,0 +1,103 @@
+        @item_order.valid?
+        expect(@item_order.errors.full_messages).to include('Postal code is invalid. Include hyphen(-)')
+      end
+      it 'shipping_area_id を選択していないと保存できないこと' do
+        @item_order.shipping_area_id = nil
+        @item_order.valid?
+        expect(@item_order.errors.full_messages).to include("Shipping area can't be blank")
+      end
```
ありがとうございます。
"nil" はから、ですね。

cony-k-m90 commented on spec/models/item_order_spec.rb:
```diff
@@ -0,0 +1,103 @@
+      it 'phone_number が空だと保存できないこと' do
+        @item_order.phone_number = ''
+        @item_order.valid?
+        expect(@item_order.errors.full_messages).to include("Phone number can't be blank", 'Phone number is invalid')
```
ありがとうございます。
不要なメッセージを削除しました。

cony-k-m90 commented on spec/models/item_order_spec.rb:
```diff
@@ -0,0 +1,103 @@
+        @item_order.valid?
+        expect(@item_order.errors.full_messages).to include('Phone number is invalid')
+      end
+      it 'phone_number が11桁以内であること' do
+        @item_order.phone_number = '0901234567'
+        @item_order.valid?
+        expect(@item_order.errors.full_messages).to include('Phone number is invalid')
+      end
```
ありがとうございます。
10桁でも保存できるテストも追記しました。

ちなみに、FactoryBotの item_orders.rb ファイルに、1つのパラメーターに対して正となる値を複数持たせることはできるのでしょうか。

cony-k-m90 committed:
商品詳細画面の編集ができる条件を修正

elseの結果に条件分岐が入ってしまっている為、修正しましょう。

cony-k-m90 commented on app/models/item_order.rb:
```diff
@@ -0,0 +1,21 @@
+    validates :city, format: { with: /\A[ぁ-んァ-ヶ一-龥々ー]+\z/, message: 'is invalid. Input full-width characters' }
+    validates :house_number, :token, :price
+    validates :phone_number, format: { with: /\A[0-9]{11}\z/ }
+  end
```
ありがとうございます。
フォームオブジェクトはモデルファイルと違うとのことで、パラメーターの扱いは気をつけなければいけないですね。
user_id , item_id を追記しました。

cony-k-m90 commented on spec/models/item_order_spec.rb:
```diff
@@ -0,0 +1,103 @@
+RSpec.describe ItemOrder, type: :model do
+  before do
+    @item_order = FactoryBot.build(:item_order)
+  end
```
ありがとうございます。
時間がかかりましたが、メンターさんに助けていただき理解できました。
下記も参考にしました。
応用Lesson5　RSpec〜PicTweetのコントローラーの単体テストコードを書こう〜

cony-k-m90 commented on spec/models/item_order_spec.rb:
```diff
@@ -0,0 +1,103 @@
+        expect(@item_order.errors.full_messages).to include('Phone number is invalid')
+      end
+    end
+  end
```
ありがとうございます。
メンターさんに助けていただき、
フォームオブジェクトに渡されたパラメーターに対して、テストコードを書くことができました。

cony-k-m90 committed:
RSpec MySQLのエラー対策

active_storageを用いたモデルの単体テストをする際に、環境によってはMySQLのエラー（MySQL client is not connected）が発生する。
今回発生してしまったため対応。

cony-k-m90 committed:
テストコード修正および追記

コードレビュー指摘箇所の修正後の対応

cony-k-m90 committed:
Rubocopでコードの最適化

cony-k-m90 commented:
# What
再度コードレビューをお願いいたします。

# Why
レビュー後指摘箇所を修正したため。


---
### 一部動作についてキャプチャーを取得しているので、ご確認ください。
- 入力に問題がある状態で購入ボタンが押されたら、購入ページに戻りエラーメッセージが表示されること
https://gyazo.com/38cf480c380c522fe10a1b4306e40f56

- 購入が完了したら、トップページまたは購入完了ページに遷移すること
https://gyazo.com/401ae86220c5aad3bd9e4a5136b038da

- ログアウト状態のユーザーは、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移すること
https://gyazo.com/b96436f7e97944d152ecf8df7e427b54
https://gyazo.com/5607c48156949d577836c50444cc6793

- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移すること
https://gyazo.com/93080202f4f0b449970777ac4c2a3dc6

- ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移すること
https://gyazo.com/d0a4c3630eccd0c5b1611b8f3b96941b

---
### 以降レビュー後追加したキャプチャーです

- 商品出品者が購入されていない商品を編集できる
https://gyazo.com/73e69ac662f6d0e189c825bc26cde2fd

- 商品出品者でも購入済みの商品は編集できない
https://gyazo.com/e332d03177f9a0191db116436aec958b

- 他のユーザーが出品した商品は編集できない
https://gyazo.com/44fc954454a6be2c556f5fc58bfc3aff

ushi10 reviewed (commented):
レビュー依頼ありがとうございます。確認いたしました。
こちら、LGTMとさせていただきます。実装お疲れさまでした。

改めて挙動を確認していただき、問題がない様子でしたらこちらのブランチを忘れずマージしましょう。
マージ後、ローカルのmasterブランチにて忘れずにプルをしましょう。 その後、新しいブランチを作成しissueを確認しながら実装に取り組みましょう。
